### PR TITLE
Merged  SystemTmpl into SystemBase 

### PR DESCRIPTION
**Pull request by fdkong**
After disccussed with @andrsd, we have the following modification to allow me bring EigenSystem from Libmesh. 

```
             +--> DisplacedSystem
             | 
SystemBase ->+--> AuxiliarySystem
             |
             +-->NonlinearSystemBase+--> NonlinearSystem
                                    |
                                    +--> NonlinearEigenSystem
```

```
SubProblem ->+--> DisplacedProblem
             |
             +-->ProblemBase+--> FEProblem
                            |
                            +--> EigenProblem
```

This PR merges SystemTmpl into SystemBase.

Refs #7762 #7654 #7398

### CODE REVIEW

**Comment by friedmud** [comment]
Acknowledging that I see this.  I won't be able to review it until tomorrow though.

**Comment by fdkong** [comment]
> Acknowledging that I see this. I won't be able to review it until tomorrow though.

Thanks @friedmud.

**Comment by fdkong** [comment]
@friedmud, @andrsd do you guys have a chance to look at this? Thanks.

File: framework/include/base/AuxiliarySystem.h [hunk 1]
```diff
@@ -55,7 +55,7 @@ class AuxiliarySystem : public SystemTempl<TransientExplicitSystem>
   virtual void jacobianSetup();
   virtual void updateActive(THREAD_ID tid);
 
-  virtual void addVariable(const std::string & var_name, const FEType & type, Real scale_factor, const std::set< SubdomainID > * const active_subdomains = NULL);
+  virtual void addVariable(const std::string & var_name, const FEType & type, Real scale_factor, const std::set< SubdomainID > * const active_subdomains = NULL) override;
```
Comment (andrsd) [inline framework/include/base/AuxiliarySystem.h]: remove the spaces inside `<...>`.

File: framework/include/base/SystemBase.h [hunk 2]
```diff
@@ -72,24 +72,42 @@ void
 dataLoad(std::istream & stream, SystemBase & system_base, void * context);
 
 /**
+ * Information about variables that will be copied
+ */
+struct VarCopyInfo {
```
Comment (andrsd) [inline framework/include/base/SystemBase.h]: `{` should be on a separate line

File: framework/include/base/SystemBase.h [hunk 3]
```diff
@@ -170,18 +192,18 @@ class SystemBase : public libMesh::ParallelObject
    * @param scale_factor the scaling factor for the variable
    * @param active_subdomains a list of subdomain ids this variable is active on
    */
-  virtual void addVariable(const std::string & var_name, const FEType & type, Real scale_factor, const std::set< SubdomainID > * const active_subdomains = NULL) = 0;
+  virtual void addVariable(const std::string & var_name, const FEType & type, Real scale_factor, const std::set< SubdomainID > * const active_subdomains = NULL);
```
Comment (andrsd) [inline framework/include/base/SystemBase.h]: `<SubdomainID>`

File: framework/include/base/SystemBase.h [hunk 4]
```diff
@@ -534,169 +411,60 @@ class SystemTempl : public SystemBase
...
-   * Remove a vector from the system with the given name.
+   * Adds a scalar variable
+   * @param var_name The name of the variable
+   * @param order The order of the variable
+   * @param scale_factor The scaling factor to be used with this scalar variable
    */
-  virtual void removeVector(const std::string & name) { _sys.remove_vector(name); }
-
-  virtual NumericVector<Number> & residualVector(Moose::KernelType /*type*/) { return *_dummy_vec; }
+  virtual void addScalarVariable(const std::string & var_name, Order order, Real scale_factor, const std::set< SubdomainID > * const active_subdomains = NULL);
```
Comment (andrsd) [inline framework/include/base/SystemBase.h]: `<SubdomainID>`

File: framework/include/base/SystemBase.h [hunk 5]
```diff
@@ -534,169 +411,60 @@ class SystemTempl : public SystemBase
...
-  virtual void init()
-  {
-  }
+  const std::vector<VariableName> & getVariableNames() const { return _vars[0].names(); };
 
-  virtual void update()
-  {
-    _sys.update();
-  }
+  virtual void computeVariables(const NumericVector<Number> & /*soln*/){ };
```
Comment (andrsd) [inline framework/include/base/SystemBase.h]: `) {`.

File: framework/src/base/SystemBase.C [hunk 6]
```diff
@@ -429,3 +433,221 @@ SystemBase::augmentSendList(std::vector<dof_id_type> & send_list)
     }
   }
 }
+
+
+/**
+ * Save the old and older solutions.
+ */
+void SystemBase::saveOldSolutions()
```
Comment (andrsd) [inline framework/src/base/SystemBase.C]: Return types on a separate line. Everywhere in this file...

File: framework/src/base/SystemBase.C [hunk 7]
```diff
@@ -429,3 +433,221 @@ SystemBase::augmentSendList(std::vector<dof_id_type> & send_list)
...
+{
+  if (hasVector(vector_name))
+    return getVector(vector_name);
+
+  NumericVector<Number> * vec = &system().add_vector(vector_name, project, type);
+  return *vec;
+}
+
+
+void SystemBase::addVariable(const std::string & var_name, const FEType & type, Real scale_factor, const std::set< SubdomainID > * const active_subdomains)
```
Comment (andrsd) [inline framework/src/base/SystemBase.C]: `<SubdomainID>`

File: framework/src/base/SystemBase.C [hunk 8]
```diff
@@ -429,3 +433,221 @@ SystemBase::augmentSendList(std::vector<dof_id_type> & send_list)
...
+  }
+  if (active_subdomains == NULL)
+    _var_map[var_num] = std::set<SubdomainID>();
+  else
+    for (std::set<SubdomainID>::iterator it = active_subdomains->begin(); it != active_subdomains->end(); ++it)
+      _var_map[var_num].insert(*it);
+}
+
+
+void SystemBase::addScalarVariable(const std::string & var_name, Order order, Real scale_factor, const std::set< SubdomainID > * const active_subdomains)
```
Comment (andrsd) [inline framework/src/base/SystemBase.C]: `<SubdomainID>`

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by fdkong** [comment]
@friedmud, could you take a look at this before we merge?

**Comment by friedmud** [comment]
Yes.

**Comment by fdkong** [comment]
> Yes.

Pinging @friedmud. Thanks.

**Comment by friedmud** [comment]
I will review this today.

**Comment by friedmud** [comment]
Reviewing

File: framework/include/base/NonlinearSystem.h [hunk 9]
```diff
@@ -457,8 +457,19 @@ class NonlinearSystem : public SystemTempl<TransientNonlinearImplicitSystem>,
...
 
+  virtual NumericVector<Number> & solution() override { return *_sys.solution; }
+
+  virtual NumericVector<Number> & solutionOld() override { return *_sys.old_local_solution; }
+
+  virtual NumericVector<Number> & solutionOlder() override { return *_sys.older_local_solution; }
+
+  virtual TransientNonlinearImplicitSystem & sys() { return _sys; }
+
+  virtual System & system() override { return _sys; }
```
Comment (friedmud) [inline framework/include/base/NonlinearSystem.h]: I gotta say - I'm not too happy about having `sys()` and `system()`... both with different return types.  I understand the reason for it... but I don't like the naming... I'll always forget which is which...

**Review by friedmud — COMMENTED** [review]
Overall: I'm ok with this.  It's just good, old fashioned PCOI.

Like I mention... I don't care for the idea of having `sys()` and `system()`... but I can cope if there really isn't a better idea (I don't have any).

I _do_ like the removal of `SystemTempl`... it was always kind of a pain... and since it was templated it means more code had to be in headers... now all of that has been moved to the .C files which is definitely better.

**Comment by friedmud** [comment]
BTW: This causes my own code to fail to compile (I have a `SystemTempl` specialization).

Just a heads up that this might not go through smoothly for everyone...

**Comment by fdkong** [comment]
> I gotta say - I'm not too happy about having sys() and system()... both with different return types. I understand the reason for it... but I don't like the naming... I'll always forget which is which...

@friedmud, How about we change  `sys()` to `transientLibmeshSystem()`, and `system()` to `libmeshSystem()`? Because we also have our own `systems` (`NonlinearSystem`, `NonlinearEigenSystem`, `AuxiliarySystem`, et.c ). This naming  distinguishes  moose `systems` from libmesh `systems`.

**Comment by fdkong** [comment]
> BTW: This causes my own code to fail to compile (I have a SystemTempl specialization).
> 
> Just a heads up that this might not go through smoothly for everyone...

What causes this problem? I  moved four functions only down to sub-classes. It should not be difficult to change any users they have a XXXsystem derived from SystemTmpl.

**Comment by fdkong** [comment]
> I gotta say - I'm not too happy about having sys() and system()... both with different return types. 

I did not see any easy way to merge those two functions together. The reason is that TransientSystem in libmesh is a class template not a class so that it can not serve as a super-class. 

>I understand the reason for it... but I don't like the naming... I'll always forget which is which...

**Comment by fdkong** [comment]
@andrsd, could you please approve my changes, if you are fine with that?

@friedmud, are we going to merge this one? Do you have any better idea on naming ?

**Comment by friedmud** [comment]
@fdkong It fails because I have a new "System" class that inherits from `SystemTempl`!

It can be fixed up... I just wanted to make you aware that anyone that has done anything similar will get a failure from this PR.

**Comment by fdkong** [comment]
> @fdkong It fails because I have a new "System" class that inherits from SystemTempl!
> 
> It can be fixed up... I just wanted to make you aware that anyone that has done anything similar will get a failure from this PR.

@friedmud, gotta.

**Review by andrsd — APPROVED** [review]

**Comment by fdkong** [comment]
> I gotta say - I'm not too happy about having sys() and system()... both with different return types. I understand the reason for it... but I don't like the naming... I'll always forget which is which...
> 
> @friedmud, How about we change sys() to transientLibmeshSystem(), and system() to libmeshSystem()? Because we also have our own systems (NonlinearSystem, NonlinearEigenSystem, AuxiliarySystem, et.c ). This naming distinguishes moose systems from libmesh systems.

@friedmud, do you like me to make this change?

**Comment by fdkong** [comment]
@permcody, @YaqiWang do you guys have any comments  on this?

**Comment by mangerij** [comment]
Just curious, do you think this PR may have changed how PETSc builds a matrix for a problem with periodic boundary conditions?

I've been on `6f016fb` and just updated today and now I'm finding some issues with incompatible vector lengths (PETSc error) when I apply PBC to some variables. Unfortunately, Ferret doesn't have a test for this sort of thing yet (a work in progress) so it never broke in CIVET.

**Comment by andrsd** [comment]
The most of the low-level PETSc calls are in libMesh, especially the ones for building matrices and vectors. We did not touch that in this PR, so I'd say this PR did not cause it...

**Pull request merged**